Convert all default CA_PATH settings to work with multiple distibutio…

sysadmind opened the pull request:
…n default locations.

This is related to :
https://pulp.plan.io/issues/2538
https://pulp.plan.io/issues/2444

This is the first step towards building a working cross distibution pulp
package. This should start to support similar cross distibution files
for other defaults in configuration settings.

This PR removes 2 tests that were based around assumptions that are no longer valid with the change in how CA_PATH is defaulted. I would like to add new tests to counter act this removal, but I'm not sure about how best to go about it. I'm very open to suggestions on testing and the make up of this change in general. I chose to use the defaults in pulp.common.constants as it keeps this change fairly small and reasonable, but it does mean that all of these packages will require pulp.common. I assume that this is a likely scenario even without this PR, but it should be noted.

dkliban commented:
ok test

sysadmind commented:
@dkliban I was finally able to get my pulp local vagrant running and all tests pass for me.

seandst commented on common/pulp/common/constants.py:
```diff
@@ -48,8 +48,12 @@
 # this is used by both platform and plugins to find the default CA path
-DEFAULT_CA_PATH = '/etc/pki/tls/certs/ca-bundle.crt'
-
+DEFAULT_CA_PATH_LIST = [
```
It might be reasonable to leave this as `DEFAULT_CA_PATH`, and set its value at run-time with `get_from_default_paths`. Otherwise, it doesn't seem useful to define this as a constant if the only useful way to access it is by calling `get_from_default_paths` on its value. Part of the motivation for this is that this constant is imported by the RPM plugin, so this change will break that plugin, maybe more.

Maybe something like this is reasonable?
```python
DEFAULT_CA_PATH_LIST = [...]
# set DEFAULT_CA_PATH based on DEFAULT_CA_PATH_LIST for backward compatibility with plugins
for path in DEFAULT_CA_PATH_LIST:
    if os.path.exists(path):
        DEFAULT_CA_PATH = path
        break
else:
    DEFAULT_CA_PATH = None
```

I've specifically avoided using `get_from_default_paths` here; just about everything in pulp imports from constants, which would create circular import problems almost anywhere the function lived.

seandst commented on bindings/test/unit/test_server.py:
```diff
@@ -210,9 +210,6 @@ def test___init___defaults(self):
         self.assertEqual(connection.verify_ssl, True)
-        self.assertEqual(connection.ca_path, server.DEFAULT_CA_PATH)
-        # 1142376 - verify default path points to a known valid file
-        self.assertEqual(server.DEFAULT_CA_PATH, '/etc/pki/tls/certs/ca-bundle.crt')
```
These assertions could probably be adapted rather than removed.

For line #213, it can maybe be changed to `self.assertTrue(connection.ca_path in server.DEFAULT_CA_PATH_LIST)`

For line #215, it seems important to have something verifying that the default CA path is set to a file that actually exists on a given platform. A new unit test covering the behavior of `get_from_default_paths` is necessary to ensure that that function does not return `None`, so that the tests appropriately fail if/when some distribution doesn't place the CA bundle in a location known to pulp.

sysadmind commented:
@seandst Thank you for taking the time to review this PR. I really appreciate it. I looked over your feedback and made some changes. I dropped the get_from_default_paths function all together in favor of your backwards compatible recommendation. I also adjusted the unit tests that I previously removed and added a new one for the pulp.common.constants module.

seandst commented on client_consumer/pulp/client/consumer/config.py:
```diff
@@ -1,7 +1,8 @@
 import socket
 
-from pulp.common.config import Config, REQUIRED, ANY, NUMBER, BOOL, OPTIONAL
+from pulp.common.config import Config, REQUIRED, ANY, NUMBER, BOOL, OPTIONAL, get_from_default_paths
```
The jenkies tests are failing because of this import, I think it's the last one left to remove.

seandst commented on common/pulp/common/constants.py:
```diff
@@ -48,7 +49,19 @@
+        DEFAULT_CA_PATH = path
+        break
+else:
+    DEFAULT_CA_PATH = None
```
It just occurred to me that it might be nice to have a comment on this, indicating that no CA path was found, and we probably need a new entry in the CA path list. What do you think?

sysadmind replied:
I agree completely. I'll write something up.

seandst commented:
I think this looks good now, but it looks like jenkins is having some network problems. Once those clear up, I'd love to see the tests pass on all platforms before merging.

seandst commented:
ok test

seandst commented on bindings/test/unit/test_server.py:
```diff
@@ -210,9 +211,9 @@ def test___init___defaults(self):
+        self.assertTrue(connection.ca_path in DEFAULT_CA_PATH_LIST)
         # 1142376 - verify default path points to a known valid file
-        self.assertEqual(server.DEFAULT_CA_PATH, '/etc/pki/tls/certs/ca-bundle.crt')
+        self.assertIsNotNone(server.DEFAULT_CA_PATH)
```
This doesn't work in el6 because it's using py2.6's unittest.
`self.assertTrue(server.DEFAULT_CA_PATH is not None)` is probably the most readable alternative.

There's another one over in `common/test/unit/common/test_constants.py` that will need to be fixed.

sysadmind commented:
@seandst Tests updated for EL6

seandst commented:
ok test

sysadmind commented:
@seandst Do you have a log of the test failures?

seandst commented:
"(295 failures / +121)" I don't think this one's on you... :)

seandst commented:
ok test

seandst commented:
This is still causing problems:

```python
from pulp.common.config import Config, REQUIRED, ANY, NUMBER, BOOL, OPTIONAL, get_from_default_paths
ImportError: cannot import name get_from_default_paths
```
in `client_consumer/pulp/client/consumer/config.py`

seandst commented:
ok test

seandst commented:
ok test

seandst commented:
ok test

seandst commented:
lookit the green check!

seandst commented:
If you haven't yet, get yourself in the authors file:
https://github.com/pulp/pulp/blob/master/AUTHORS

While you're in there, maybe add me, too. :D (just kidding)

sysadmind commented:
@seandst Thanks again for the review and your help to get this merged. I did add myself in a previous commit to the authors file. Much appreciated! :tada:
